packages/portalnetwork/src/subprotocols/protocol.ts

```diff
@@ -29,23 +28,64 @@ import {
   RequestCode,
   NodeLookup,
   StateNetworkRoutingTable,
+  encodeWithVariantPrefix,
+  INewRequest,
+  ContentRequest,
 } from '../index.js'
-import { bigIntToHex } from '@ethereumjs/util'
-export abstract class BaseProtocol {
+import { EventEmitter } from 'events'
+
+export interface IProtocolOpts {
+  sendMessage: (
+    enr: ENR | string,
+    payload: Buffer,
+    protocolId: ProtocolId,
+    utpMessage?: boolean
+  ) => Promise<Buffer>
+  sendResponse: (src: INodeAddress, requestId: bigint, payload: Uint8Array) => Promise<void>
+  findEnr: (nodeId: string) => ENR | undefined
+  put: (contentKey: string, content: string) => void
+  get: (contentKey: string) => Promise<string>
+  handleNewRequest: (request: INewRequest) => Promise<ContentRequest>
+  prune: (radius: bigint) => Promise<void>
+  ENR: ENR
+  nodeRadius?: bigint
+  metrics?: PortalNetworkMetrics
+}
+export abstract class BaseProtocol extends EventEmitter {
   public routingTable: PortalNetworkRoutingTable | StateNetworkRoutingTable
-  protected metrics: PortalNetworkMetrics | undefined
+  public metrics: PortalNetworkMetrics | undefined
   private nodeRadius: bigint
   private checkIndex: number
-  protected abstract logger: Debugger
+  abstract logger: Debugger
   abstract protocolId: ProtocolId
   abstract protocolName: string
-  public client: PortalNetwork
-  constructor(client: PortalNetwork, nodeRadius?: bigint, metrics?: PortalNetworkMetrics) {
-    this.client = client
+  public ENR: ENR
+  handleNewRequest: (request: INewRequest) => Promise<ContentRequest>
```

Comment:
I'm not really sure how I feel about this one.  While I sort of get why you pulled everything out like this, it still obscures the implementation details of all these methods and makes them hard to find.  I guess I could go either way with it at the end of the day if you have a compelling reason for abstracting all these methods out but it feels weird to me to have a bunch of methods with no implementation details available within the protocol.  The main case I can think of for doing it this way is if we're going to build separate client implementations that interact with these protocols and I don't think that'll ever be the case (unless we completely rip out the guts of the client and replace the whole thing with a libp2p client - for example).  That, to be honest and despite my preference, is probably a long way off.  If it were just me, I'd keep everything as it was before so we have access to the full client as needed the next time we're building out a subprotocol.  I dunno, will have to think more on it.

Reply:
OK -- I can live with rolling back that part.  Not really in the scope of this PR anyway.